File: app/src/main/java/jp/hazuki/yuzubrowser/action/item/WebScrollSingleAction.java
package jp.hazuki.yuzubrowser.action.item;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.res.Resources;
import android.hardware.SensorManager;
import android.os.Parcel;
import android.os.Parcelable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewConfiguration;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.IOException;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.action.SingleAction;
import jp.hazuki.yuzubrowser.action.view.ActionActivity;
import jp.hazuki.yuzubrowser.utils.ArrayUtils;
import jp.hazuki.yuzubrowser.utils.DisplayUtils;
import jp.hazuki.yuzubrowser.utils.app.StartActivityInfo;
import jp.hazuki.yuzubrowser.webkit.CustomWebView;

public class WebScrollSingleAction extends SingleAction implements Parcelable {
    private static final String FIELD_NAME_TYPE = "0";
    private static final String FIELD_NAME_X = "1";
    private static final String FIELD_NAME_Y = "2";
    public static final int TYPE_FAST = 0;
    public static final int TYPE_FLING = 1;
    private int mType = TYPE_FLING;
    private int mX, mY;

    public WebScrollSingleAction(int id, JsonParser parser) throws IOException {
        super(id);

        if (parser != null) {
            if (parser.nextToken() != JsonToken.START_OBJECT) return;
            while (parser.nextToken() != JsonToken.END_OBJECT) {
                if (parser.getCurrentToken() != JsonToken.FIELD_NAME) return;
                if (FIELD_NAME_TYPE.equals(parser.getCurrentName())) {
                    if (parser.nextToken() != JsonToken.VALUE_NUMBER_INT) return;
                    mType = parser.getIntValue();
                    continue;
                }
                if (FIELD_NAME_X.equals(parser.getCurrentName())) {
                    if (parser.nextToken() != JsonToken.VALUE_NUMBER_INT) return;
                    mX = parser.getIntValue();
                    continue;
                }
                if (FIELD_NAME_Y.equals(parser.getCurrentName())) {
                    if (parser.nextToken() != JsonToken.VALUE_NUMBER_INT) return;
                    mY = parser.getIntValue();
                    continue;
                }
                parser.skipChildren();
            }
        }
    }

    @Override
    public void writeIdAndData(JsonGenerator generator) throws IOException {
        generator.writeNumber(id);
        generator.writeStartObject();
        generator.writeNumberField(FIELD_NAME_TYPE, mType);
        generator.writeNumberField(FIELD_NAME_X, mX);
        generator.writeNumberField(FIELD_NAME_Y, mY);
        generator.writeEndObject();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(id);
        dest.writeInt(mType);
        dest.writeInt(mX);
        dest.writeInt(mY);
    }

    protected WebScrollSingleAction(Parcel source) {
        super(source.readInt());
        mType = source.readInt();
        mX = source.readInt();
        mY = source.readInt();
    }

    public static final Creator<WebScrollSingleAction> CREATOR = new Creator<WebScrollSingleAction>() {
        @Override
        public WebScrollSingleAction createFromParcel(Parcel source) {
            return new WebScrollSingleAction(source);
        }

        @Override
        public WebScrollSingleAction[] newArray(int size) {
            return new WebScrollSingleAction[size];
        }
    };

    @Override
    public StartActivityInfo showMainPreference(ActionActivity context) {
        return showSubPreference(context);
    }

    @Override
    public StartActivityInfo showSubPreference(final ActionActivity context) {
        View view = LayoutInflater.from(context).inflate(R.layout.action_web_scroll_setting, null);
        final Spinner typeSpinner = (Spinner) view.findViewById(R.id.typeSpinner);
        final EditText editTextX = (EditText) view.findViewById(R.id.editTextX);
        final EditText editTextY = (EditText) view.findViewById(R.id.editTextY);

        Resources res = context.getResources();
        ArrayAdapter<String> adapter = new ArrayAdapter<>(context, android.R.layout.simple_spinner_item, res.getStringArray(R.array.action_web_scroll_type_list));
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        typeSpinner.setAdapter(adapter);
        final int values[] = res.getIntArray(R.array.action_web_scroll_type_values);
        int current = ArrayUtils.findIndexOfValue(mType, values);
        if (current < 0)
            current = TYPE_FLING;
        typeSpinner.setSelection(current);

        editTextX.setText(String.valueOf(mX));
        editTextY.setText(String.valueOf(mY));

        new AlertDialog.Builder(context)
                .setTitle(R.string.action_settings)
                .setView(view)
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        mType = values[typeSpinner.getSelectedItemPosition()];

                        int x = 0;
                        int y = 0;

                        try {
                            x = Integer.parseInt(editTextX.getText().toString());
                        } catch (NumberFormatException e) {
                            e.printStackTrace();
                        }

                        try {
                            y = Integer.parseInt(editTextY.getText().toString());
                        } catch (NumberFormatException e) {
                            e.printStackTrace();
                        }

                        if (x == 0 && y == 0) {
                            Toast.makeText(context.getApplicationContext(), R.string.action_web_scroll_x_y_zero, Toast.LENGTH_SHORT).show();
                            showSubPreference(context);
                            return;
                        }

                        mX = x;
                        mY = y;
                    }
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();

        return null;
    }

    public int getIconResourceId() {
        if (mX > 0) {
            if (mY > 0)
                return R.drawable.ic_arrow_down_right_white_24dp;
            else if (mY < 0)
                return R.drawable.ic_arrow_up_right_white_24px;
            else
                return R.drawable.ic_arrow_forward_white_24dp;
        } else if (mX < 0) {
            if (mY > 0)
                return R.drawable.ic_arrow_down_left_24dp;
            else if (mY < 0)
                return R.drawable.ic_arrow_up_left_white_24px;
            else
                return R.drawable.ic_arrow_back_white_24dp;
        } else {
            if (mY > 0)
                return R.drawable.ic_arrow_downward_white_24dp;
            else if (mY < 0)
                return R.drawable.ic_arrow_upward_white_24dp;
            else
                return -1;
        }
    }

    public void scrollWebView(Context context, CustomWebView web) {
        switch (mType) {
            case TYPE_FAST:
                web.scrollBy(mX, mY);
                break;
            case TYPE_FLING:
                web.flingScroll(makeFlingValue(context, mX), makeFlingValue(context, mY));
                break;
            default:
                throw new RuntimeException("Unknown type : " + mType);
        }
    }

    private int makeFlingValue(Context context, int d) {
        final float DECELERATION_RATE = (float) (Math.log(0.78) / Math.log(0.9));
        final float PHYSICAL_COEF = SensorManager.GRAVITY_EARTH * 39.37f * DisplayUtils.getDensity(context) * 160.0f * 0.84f;
        final float FLING_FLICTION = ViewConfiguration.getScrollFriction();
        return (int) (Integer.signum(d) * Math.round(Math.exp(Math.log(Math.abs(d) / (FLING_FLICTION * PHYSICAL_COEF)) / DECELERATION_RATE * (DECELERATION_RATE - 1.0)) / 0.35f * (FLING_FLICTION * PHYSICAL_COEF)));
    }
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/action/item/WebScrollSingleAction.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.action.item

import android.app.AlertDialog
import android.content.Context
import android.hardware.SensorManager
import android.os.Parcel
import android.os.Parcelable
import android.view.View
import android.view.ViewConfiguration
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.Spinner
import android.widget.Toast
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import jp.hazuki.yuzubrowser.core.utility.extensions.density
import jp.hazuki.yuzubrowser.core.utility.utils.ArrayUtils
import jp.hazuki.yuzubrowser.legacy.R
import jp.hazuki.yuzubrowser.legacy.action.SingleAction
import jp.hazuki.yuzubrowser.legacy.action.view.ActionActivity
import jp.hazuki.yuzubrowser.ui.app.StartActivityInfo
import jp.hazuki.yuzubrowser.webview.CustomWebView
import java.io.IOException

class WebScrollSingleAction : SingleAction, Parcelable {
    private var mType = TYPE_FLING
    private var mX: Int = 0
    private var mY: Int = 0

    val iconResourceId: Int
        get() = when {
            mX > 0 -> when {
                mY > 0 -> R.drawable.ic_arrow_down_right_white_24dp
                mY < 0 -> R.drawable.ic_arrow_up_right_white_24px
                else -> R.drawable.ic_arrow_forward_white_24dp
            }
            mX < 0 -> when {
                mY > 0 -> R.drawable.ic_arrow_down_left_24dp
                mY < 0 -> R.drawable.ic_arrow_up_left_white_24px
                else -> R.drawable.ic_arrow_back_white_24dp
            }
            else -> when {
                mY > 0 -> R.drawable.ic_arrow_downward_white_24dp
                mY < 0 -> R.drawable.ic_arrow_upward_white_24dp
                else -> -1
            }
        }

    @Throws(IOException::class)
    constructor(id: Int, reader: JsonReader?) : super(id) {

        if (reader != null) {
            if (reader.peek() != JsonReader.Token.BEGIN_OBJECT) return
            reader.beginObject()
            while (reader.hasNext()) {
                if (reader.peek() != JsonReader.Token.NAME) return
                when (reader.nextName()) {
                    FIELD_NAME_TYPE -> mType = reader.nextInt()
                    FIELD_NAME_X -> mX = reader.nextInt()
                    FIELD_NAME_Y -> mY = reader.nextInt()
                    else -> reader.skipValue()
                }
            }
            reader.endObject()
        }
    }

    @Throws(IOException::class)
    override fun writeIdAndData(writer: JsonWriter) {
        writer.value(id)
        writer.beginObject()
        writer.name(FIELD_NAME_TYPE)
        writer.value(mType)
        writer.name(FIELD_NAME_X)
        writer.value(mX)
        writer.name(FIELD_NAME_Y)
        writer.value(mY)
        writer.endObject()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(id)
        dest.writeInt(mType)
        dest.writeInt(mX)
        dest.writeInt(mY)
    }

    private constructor(source: Parcel) : super(source.readInt()) {
        mType = source.readInt()
        mX = source.readInt()
        mY = source.readInt()
    }

    override fun showMainPreference(context: ActionActivity): StartActivityInfo? {
        return showSubPreference(context)
    }

    override fun showSubPreference(context: ActionActivity): StartActivityInfo? {
        val view = View.inflate(context, R.layout.action_web_scroll_setting, null)
        val typeSpinner = view.findViewById<Spinner>(R.id.typeSpinner)
        val editTextX = view.findViewById<EditText>(R.id.editTextX)
        val editTextY = view.findViewById<EditText>(R.id.editTextY)

        val res = context.resources
        val adapter = ArrayAdapter(context, android.R.layout.simple_spinner_item, res.getStringArray(R.array.action_web_scroll_type_list))
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        typeSpinner.adapter = adapter
        val values = res.getIntArray(R.array.action_web_scroll_type_values)
        var current = ArrayUtils.findIndexOfValue(mType, values)
        if (current < 0)
            current = TYPE_FLING
        typeSpinner.setSelection(current)

        editTextX.setText(mX.toString())
        editTextY.setText(mY.toString())

        AlertDialog.Builder(context)
                .setTitle(R.string.action_settings)
                .setView(view)
                .setPositiveButton(android.R.string.ok) { _, _ ->
                    mType = values[typeSpinner.selectedItemPosition]

                    var x = 0
                    var y = 0

                    try {
                        x = Integer.parseInt(editTextX.text.toString())
                    } catch (e: NumberFormatException) {
                        e.printStackTrace()
                    }

                    try {
                        y = Integer.parseInt(editTextY.text.toString())
                    } catch (e: NumberFormatException) {
                        e.printStackTrace()
                    }

                    if (x == 0 && y == 0) {
                        Toast.makeText(context.applicationContext, R.string.action_web_scroll_x_y_zero, Toast.LENGTH_SHORT).show()
                        showSubPreference(context)
                        return@setPositiveButton
                    }

                    mX = x
                    mY = y
                }
                .setNegativeButton(android.R.string.cancel, null)
                .show()

        return null
    }

    fun scrollWebView(context: Context, web: CustomWebView) {
        when (mType) {
            TYPE_FAST -> web.scrollBy(mX, mY)
            TYPE_FLING -> web.flingScroll(makeFlingValue(context, mX), makeFlingValue(context, mY))
            else -> throw RuntimeException("Unknown type : " + mType)
        }
    }

    private fun makeFlingValue(context: Context, d: Int): Int {
        val decelerationLate = (Math.log(0.78) / Math.log(0.9)).toFloat()
        val physicalCoef = SensorManager.GRAVITY_EARTH * 39.37f * context.density * 160.0f * 0.84f
        val flingFliction = ViewConfiguration.getScrollFriction()
        return (Integer.signum(d) * Math.round(Math.exp(Math.log((Math.abs(d) / (flingFliction * physicalCoef)).toDouble()) / decelerationLate * (decelerationLate - 1.0)) / 0.35f * (flingFliction * physicalCoef))).toInt()
    }

    companion object {
        private const val FIELD_NAME_TYPE = "0"
        private const val FIELD_NAME_X = "1"
        private const val FIELD_NAME_Y = "2"
        private const val TYPE_FAST = 0
        private const val TYPE_FLING = 1

        @JvmField
        val CREATOR: Parcelable.Creator<WebScrollSingleAction> = object : Parcelable.Creator<WebScrollSingleAction> {
            override fun createFromParcel(source: Parcel): WebScrollSingleAction {
                return WebScrollSingleAction(source)
            }

            override fun newArray(size: Int): Array<WebScrollSingleAction?> {
                return arrayOfNulls(size)
            }
        }
    }
}
